core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -526,22 +526,20 @@ public boolean apply(@Nullable RecoveryStatus status) {
                         // closing the shard will also cancel any ongoing recovery.
                         indexService.removeShard(shardRouting.id(), "removing shard (recovery source node changed)");
                         shardHasBeenRemoved = true;
-
                     }
                 }
-                if (shardHasBeenRemoved == false && (shardRouting.equals(indexShard.routingEntry()) == false || shardRouting.version() > indexShard.routingEntry().version())) {
-                    if (shardRouting.primary() && indexShard.routingEntry().primary() == false && shardRouting.initializing() && indexShard.allowsPrimaryPromotion() == false) {
-                        logger.debug("{} reinitialize shard on primary promotion", indexShard.shardId());
-                        indexService.removeShard(shardId, "promoted to primary");
-                    } else {
-                        // if we happen to remove the shardRouting by id above we don't need to jump in here!
-                        indexShard.updateRoutingEntry(shardRouting, event.state().blocks().disableStatePersistence() == false);
-                    }
+
+                if (shardHasBeenRemoved == false) {
+                    // shadow replicas do not support primary promotion. The master would reinitialize the shard, giving it a new allocation, meaning we should be there.
+                    // nocommit: double check check this
```

Comment:
yeah, my reasoning (which I wanted double checked) is that the master always generates a new allocation id in this case and thus the shard is cleaned by the allocation id check. No need to have special shadow replica handling (but I did add an assert and a comment).